utils/zerovec/src/map/vecs.rs

```diff
@@ -0,0 +1,126 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::ule::*;
+use crate::VarZeroVec;
+use crate::ZeroVec;
+use std::mem;
+
+/// Trait abstracting over [`ZeroVec`] and [`VarZeroVec`], for use in [`ZeroMap`]. You
+/// should not be implementing or calling this trait directly.
+pub trait ZeroVecLike<'a, T> {
+    /// The type received by `Self::binary_search()`
+    type NeedleType: ?Sized;
+    /// The type returned by `Self::get()`
+    type GetType: ?Sized;
+    /// Search for a key in a sorted vector, returns `Ok(index)` if found,
+    /// returns `Err(insert_index)` if not found, where `insert_index` is the
+    /// index where it should be inserted to maintain sort order.
+    fn binary_search(&self, k: &Self::NeedleType) -> Result<usize, usize>;
+    /// Get element at `index`
+    fn get(&self, index: usize) -> Option<&Self::GetType>;
+    /// Insert an element at `index`
+    fn insert(&mut self, index: usize, value: T);
+    /// Remove the element at `index` (panicking if nonexistant)
+    fn remove(&mut self, index: usize) -> T;
+    /// Replace the element at `index` with another one, returning the old element
+    fn replace(&mut self, index: usize, value: T) -> T;
+    /// Push an element to the end of this vector
+    fn push(&mut self, value: T);
```

Comment:
I think the problem is that you have to do two conversions for every patching operation, then. That can be pretty expensive. `ZeroMap`'s owned variant is _basically_ a `LiteMap` anyway, except it's two vectors instead of one. We're going to be directly mutating `ZeroVec`s for patching so I don't see why we shouldn't do the same for `LiteMap`.

The way I see it is that with `ZeroMap` we won't need `LiteMap` at all, because it _is_ `LiteMap`.

Also, regarding ULEs leaking out, ULEs only leak out in `.get()` and `.iter()`, which are immutable operations. The mutation doesn't make this worse.

Reply:
> I think the problem is that you have to do two conversions for every patching operation, then. That can be pretty expensive.

Do we care about the performance of patching operations?  We know they need to *work*, but do they need to be *efficient*?

And, as you said, since ZeroMap is basically a LiteMap, perhaps the conversion from LiteMap to ZeroMap is simply taking ownership of the LiteMap vectors.  So you need to convert from bytes to vectors, but not back the other way, if you use LiteMap as the mutable type.

Basically, anything of the form `.as_mut(&mut self) -> Vec<T>` locks us in to supporting `Vec<T>`.  I don't think we should be exposing that internal implementation detail.  Maybe we would one day want to change `Vec<T>` to `SmallVec<T>` or something else entirely.

> The way I see it is that with ZeroMap we won't need LiteMap at all, because it is LiteMap.

LiteMap is more general because it doesn't have all the weird restrictions that ZeroMap puts on the types it uses.

> Also, regarding ULEs leaking out, ULEs only leak out in .get() and .iter(), which are immutable operations. The mutation doesn't make this worse.

Acknowledged

Reply:
> And, as you said, since ZeroMap is basically a LiteMap, perhaps the conversion from LiteMap to ZeroMap is simply taking ownership of the LiteMap vectors. So you need to convert from bytes to vectors, but not back the other way, if you use LiteMap as the mutable type.


It's algorithmically a LiteMap, but not actually data wise, because LiteMap is a single buffer.

> Basically, anything of the form .as_mut(&mut self) -> &mut Vec<T> locks us in to supporting Vec<T>. I don't think we should be exposing that internal implementation detail. Maybe we would one day want to change Vec<T> to SmallVec<T> or something else entirely.


That implementation detail could be simply exposed as `pub(crate)` so that `ZeroMap` can use it but not others. I think it's useful independently, and if we change the backing storage we can make a breaking change to `zerovec` without breaking the serialization format. I think that would be fine.

We could also add replace/push/etc as methods on `ZeroVec` but I feel like, for now, the `&mut Vec<T>` methods are pretty nice.

Reply:
> It's algorithmically a LiteMap, but not actually data wise, because LiteMap is a single buffer.

We own LiteMap, so changing LiteMap to fit this use case is not a problem.  Indeed, that is exactly what my expectation was coming into this.

> That implementation detail could be simply exposed as pub(crate) so that ZeroMap can use it but not others.

I thought the value proposition you stated for `.as_mut(&mut self) -> &mut Vec<T>` was so that external clients could use it to patch their data.

Reply:
> I thought the value proposition you stated for `.as_mut(&mut self) -> &mut Vec<T>` was so that external clients could use it to patch their data.

It is! I'm saying that we can instead add `push()`, `replace()`, `insert()` etc on `ZeroVec` instead, if we need. I find this the easiest way to move forward for now, and we can figure this out before we 1.0 zerovec.

The only thing that would lock us into is that `ZeroVec` would be a permanently copy-on-write-enabled structure, but that's the goal -- we can always expose `ZeroVecBorrowed` and `SliceComponents` (renamed, probably) for the purely-readonly-borrowed version.

Reply:
As discussed, I'm okay now with mutating methods so long as they don't lock us into a specific internal representation.